### PR TITLE
Threads are hard

### PR DESCRIPTION
**Pull request by kripken**
More fixes for #1376

Includes a proper test, which did not pass before and now does.

### CODE REVIEW

**Comment by kripken** [comment]
Added a refactoring to give workers a direct pointer to the pool, avoiding them using a global variable which was less nice. This should also fix an issue noticed about destruction order of that global in #1376

File: auto_update_tests.py [hunk 1]
```diff
@@ -184,6 +184,8 @@
            src, '-c', '-o', 'example.o',
            '-Isrc', '-g', '-L' + libdir, '-pthread']
   print 'build: ', ' '.join(extra)
+  if src.endswith('.cpp'):
+    extra += ['-std=c++11']
```
Comment (dschuff) [inline auto_update_tests.py]: I thought clang already defaulted to C++11?
(actually I think its gnu++11, but that doesn't matter, does it?)

Comment (kripken) [reply auto_update_tests.py]: Depends on the version of clang and gcc. When I run locally in Ubuntu LTS using gcc, it needs that flag.

**Comment by kripken** [comment]
Yeah, I'll document this better in the final commit here, good point.

About the threadPool being global or not, there are advantages both ways, yeah. I prefer a single global one for a library like binaryen, since threads are a single global resource in a sense. Even if you create multiple modules and optimize them in parallel, you usually only want a number of threads active that is equal to the number of cores, and not potentially the number of cores times the number of modules.

**Comment by binji** [comment]
It seems much of the complexity here comes from having `ThreadPool` auto-initialization on `ThreadPool::get`. Even if you kept a global `ThreadPool`, you could probably simplify a lot by requiring the user to explicitly initialize the global `ThreadPool` object.

**Comment by kripken** [comment]
True, but after this PR I think most of that complexity is gone. The one place it's left is that we lock in the `get()` method, which doesn't seem so bad to me.

**Comment by kripken** [comment]
Let's merge this in, as it is confirmed to fix all known problems here. We can do further improvements later.

**Pull request merged**